### PR TITLE
Publish snapshots to maven

### PR DESCRIPTION
**Pull request by peternied**
### Description
Adds a workflow to publish snapshots to the OpenSearch-Project maven repository when updates are pushed to main, 1.* branches, or 2.* branches.

Following the example from https://github.com/opensearch-project/opensearch-sdk-java

### Issues Resolved
- Related https://github.com/opensearch-project/opensearch-build/issues/3199

### Testing
Note; cannot test the actual workflow, but I've confirmed the artifacts produced match what we need to consume

```
% find ~/.m2/* | grep security                                                                                                                                                                        ~/git/security
/home/petern/.m2/repository/org/opensearch/plugin/opensearch-security
/home/petern/.m2/repository/org/opensearch/plugin/opensearch-security/maven-metadata-local.xml
/home/petern/.m2/repository/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT
/home/petern/.m2/repository/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/maven-metadata-local.xml
/home/petern/.m2/repository/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-SNAPSHOT.zip
/home/petern/.m2/repository/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/opensearch-security-3.0.0.0-SNAPSHOT.pom
```

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by peternied** [comment]
@gaiksaya We need our repository setup so we can retrieve the credentials for access Sonotype, can you help me get that done?

**Review by DarshitChanpura — DISMISSED** [review]

**Comment by prudhvigodithi** [comment]
Hey @peternied this PR should also consider the pluginZips to be published under `org.opensearch.plugin` group.
`./gradlew assemble --no-daemon --refresh-dependencies -DskipTests=true` and `./gradlew publishPluginZipPublicationToZipStagingRepository` would create a folder `build/local-staging-repo/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/` locally with maven zips of security plugin
```
-rw-r--r--  1 pgodithi  staff       796 Feb 14 13:36 maven-metadata.xml
-rw-r--r--  1 pgodithi  staff        40 Feb 14 13:36 maven-metadata.xml.sha1
-rw-r--r--  1 pgodithi  staff        32 Feb 14 13:36 maven-metadata.xml.md5
-rw-r--r--  1 pgodithi  staff        64 Feb 14 13:36 maven-metadata.xml.sha256
-rw-r--r--  1 pgodithi  staff       128 Feb 14 13:36 maven-metadata.xml.sha512
-rw-r--r--  1 pgodithi  staff  44192967 Feb 14 13:36 opensearch-security-3.0.0.0-20230214.213641-1.zip
-rw-r--r--  1 pgodithi  staff        40 Feb 14 13:36 opensearch-security-3.0.0.0-20230214.213641-1.zip.sha1
-rw-r--r--  1 pgodithi  staff        32 Feb 14 13:36 opensearch-security-3.0.0.0-20230214.213641-1.zip.md5
-rw-r--r--  1 pgodithi  staff        64 Feb 14 13:36 opensearch-security-3.0.0.0-20230214.213641-1.zip.sha256
-rw-r--r--  1 pgodithi  staff       128 Feb 14 13:36 opensearch-security-3.0.0.0-20230214.213641-1.zip.sha512
-rw-r--r--  1 pgodithi  staff      1037 Feb 14 13:36 opensearch-security-3.0.0.0-20230214.213641-1.pom
-rw-r--r--  1 pgodithi  staff        40 Feb 14 13:36 opensearch-security-3.0.0.0-20230214.213641-1.pom.sha1
-rw-r--r--  1 pgodithi  staff        32 Feb 14 13:36 opensearch-security-3.0.0.0-20230214.213641-1.pom.md5
-rw-r--r--  1 pgodithi  staff        64 Feb 14 13:36 opensearch-security-3.0.0.0-20230214.213641-1.pom.sha256
-rw-r--r--  1 pgodithi  staff       128 Feb 14 13:36 opensearch-security-3.0.0.0-20230214.213641-1.pom.sha512
```
we should also ship them to the maven so that the plugin zips are also published via this workflow.
@gaiksaya @bbarani 
Thank you

**Review by prudhvigodithi — CHANGES_REQUESTED** [review]
https://github.com/opensearch-project/security/pull/2438#issuecomment-1430428613

**Comment by peternied** [comment]
@prudhvigodithi I've updated the description with that output

**Comment by gaiksaya** [comment]
So I cloned your fork @peternied and saw that the task differs repo to repo based on how publications are configured. For this repo seems like it is `publishMavenPublicationToSnapshotsRepository` 
The maven artifacts need to include zip (along with associated poms and shas) and jars as well. The current task `publishMavenPublicationToSnapshotsRepository` I believe will only take care of jars (previous PR output that @peternied posted). We need to find a way to tell that task about the zip location as well. Maybe something like this https://stackoverflow.com/questions/41346425/publish-a-zip-file-to-nexus-maven-with-gradle

File: .github/workflows/maven-publish.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        with:
+          role-to-assume: ${{ secrets.PUBLISH_SNAPSHOTS_ROLE }}
+          aws-region: us-east-1
+      - name: publish snapshots to maven
+        run: |
+          export SONATYPE_USERNAME=$(aws secretsmanager get-secret-value --secret-id maven-snapshots-username --query SecretString --output text)
+          export SONATYPE_PASSWORD=$(aws secretsmanager get-secret-value --secret-id maven-snapshots-password --query SecretString --output text)
+          echo "::add-mask::$SONATYPE_USERNAME"
+          echo "::add-mask::$SONATYPE_PASSWORD"
+          ./gradlew publishMavenJavaPublicationToSnapshotsRepository
```
Comment (peternied) [inline .github/workflows/maven-publish.yml:42]: This is the wrong publishing task, it creates the jars for upload, not the zip.

Comment (gaiksaya) [reply .github/workflows/maven-publish.yml:42]: Exactly. This is how it was handled before https://github.com/opensearch-project/opensearch-build/blob/main/scripts/default/opensearch/build.sh#L73-L83
And then maven cli was called via this script to publish everything https://github.com/opensearch-project/opensearch-build/blob/main/publish/publish-snapshot.sh
We need to figure out a way to do that directly via publishing task now.

**Comment by peternied** [comment]
@prudhvigodithi @gaiksaya I've fix the publish task and wrote the test process into the description.

**Comment by gaiksaya** [comment]
@peternied The jars are missing from that output. I believe we want all of these https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/

**Comment by gaiksaya** [comment]
Update: We do not publish jars for security plugin. In that case we can also remove https://github.com/opensearch-project/security/blob/main/build.gradle#L64-L67
Adding @prudhvigodithi for more context

**Comment by prudhvigodithi** [comment]
To add more context security plugin at preset is used only via zips (may be this could change in future), example https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/opensearch-security/3.0.0.0-SNAPSHOT/, the folder all contains zips. For zips publish we already have a reserved groupID `org.opensearch.plugin` https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/opensearch-security/3.0.0.0-SNAPSHOT/  which is taken care by the plugin [opensearch.pluginzip](https://github.com/opensearch-project/security/blob/main/build.gradle#LL70C15-L70C35). 

So long story short for security plugin lets only focus on shipping the files part of  `org/opensearch/plugin` folder to maven using GH workflows.
@peternied

**Review by gaiksaya — APPROVED** [review]
Will comment back once the role is added to the secrets!

**Comment by gaiksaya** [comment]
> Will comment back once the role is added to the secrets!
> @gaiksaya We need our repository setup so we can retrieve the credentials for access Sonotype, can you help me get that done?

The role has been added to the secrets. Thanks!

**Comment by peternied** [comment]
@prudhvigodithi I noticed that you have requested changes, but it's not entirely clear to me what changes you would like to see. Could you please provide more specific feedback on what needs to be modified or improved so that I can make the necessary adjustments to the code?

**Comment by prudhvigodithi** [comment]
Hey @peternied I have requested changes, before we considered the the task that picks up the files from folder `/org/opensearch/plugin/`, but not looks good now.
Thank you

**Review by prudhvigodithi — APPROVED** [review]

**Review by DarshitChanpura — APPROVED** [review]

**Comment by DarshitChanpura** [comment]
@peternied Please merge this if you aren't making any more changes.

**Comment by gaiksaya** [comment]
@peternied Give it a try? https://github.com/opensearch-project/security/actions/workflows/maven-publish.yml

**Comment by peternied** [comment]
@gaiksaya Nice, it works!  I've got some cleanup - the multiplatform build is entirely unneeded.

https://github.com/opensearch-project/security/actions/runs/4196975415/jobs/7278687473
